src/main.rs

```diff
@@ -70,6 +92,116 @@ impl BFInterp {
+            Right(op) => match op {
+                LocalOperand { .. } => todo!(),
+                ConstantOperand(con_op) => match con_op.as_ref() {
+                    Constant::GlobalReference { name, .. } => match name {
+                        Name(n) => {
+                            if n.as_str() == "printf" {
+                                self.exe_printf(call);
+                            }
+                        }
+                        Number(_) => todo!(),
+                    },
+                    _ => todo!(),
+                },
+                MetadataOperand => todo!(),
+            },
+        }
+    }
+
+    fn exe_printf(&mut self, call: &Call) {
+        let constant = match &call.arguments[0].0 {
+            LocalOperand { .. } => todo!(),
+            ConstantOperand(const_op) => match const_op.as_ref() {
+                Constant::GetElementPtr(GetElementPtr { address, .. }) => match address.as_ref() {
+                    Constant::GlobalReference { name, .. } => self.gl_vars.get(name).unwrap(),
+                    _ => todo!(),
+                },
+                _ => todo!(),
+            },
+            MetadataOperand => todo!(),
+        };
+
+        let mut string = String::new();
+        let mut arg_it = call.arguments[1..].into_iter();
+        match constant {
+            Constant::Array { elements, .. } => {
+                let mut i = 0;
+                while i < elements.len() {
+                    match elements[i].as_ref() {
+                        Constant::Int { value, .. } => {
+                            let ch = *value as u8 as char;
+                            if ch == '%' {
+                                string.push_str(
+                                    self.arg_to_string(&arg_it.next().unwrap().0).as_str(),
+                                );
+                                i += 2;
+                                continue;
+                            } else if value == &0u64 {
+                                i += 1;
+                                continue;
+                            }
+                            string.push(ch);
+                        }
+                        _ => todo!(),
+                    }
+                    i += 1;
+                }
+            }
+            _ => todo!(),
+        }
+
+        println!("{}", string);
+    }
+
+    fn arg_to_string(&self, arg: &Operand) -> String {
+        match arg {
+            LocalOperand { name, .. } => self.mem.get(&name).unwrap().to_string(),
+            ConstantOperand(con_op) => match con_op.as_ref() {
+                Constant::Int { value, .. } => value.to_string(),
+                Constant::GetElementPtr(GetElementPtr { address, .. }) => match address.as_ref() {
+                    Constant::GlobalReference { name, .. } => {
+                        match self.gl_vars.get(name).unwrap() {
+                            Constant::Array { elements, .. } => {
```

Comment:
I'm wondering it we should put this array to string conversion into its own function, as we might need to do this conversion more often later on. We could possibly even use this to store the global constant string arrays as actual strings (though the might affect performance a bit since we then iterate over them again if they are c format strings). Thoughts?

Reply:
I'm happy to put it into another function if you think that's better. 
I should probably say as a different point that my method assumes that the global variable in question is a string which might cause problems but if this is just for testing anyway and external function calls are going to be done differently anyway then it might not matter too much right now.

Reply:
Yes, the function can assume that the given array is a string. Maybe we can add a `debug_assert!` somewhere to make sure that we only pass arrays that can be strings into the function.

Reply:
I added an assert! instead of a debug_assert! as the debug_assert! didn't seem to be executed when I ran cargo test.

Reply:
`debug_assert!` is only executed in debug mode. So when you run cargo test in release mode it doesn't execute.

Reply:
Would it be better to use debug_assert! or is assert! fine?

Reply:
Since `lang_tester` runs in release mode, I guess it's better to leave it as `assert`, though there is an argument to be made whether `lang_tester` should run in debug mode. Let's leave it as is.